llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp

```diff
@@ -1220,7 +1236,7 @@ Value *AMDGPUCodeGenPrepareImpl::expandDivRem24(IRBuilder<> &Builder,
   // If Num bits <= 24, assume 0 signbits.
   unsigned AtLeast = (SSBits <= 24) ? 0 : (SSBits - 24 + IsSigned);
   int DivBits = getDivNumBits(I, Num, Den, AtLeast, IsSigned);
-  if (DivBits == -1)
+  if (DivBits == -1 || DivBits > 24)
```

Comment:
PLEASE take a moment to simplify things by cleaning up the API of `getDivNumBits`, instead of just adding extra checks.

`getDivNumBits` should return unsigned. There is no need for the special -1 return value, since returning bitwidth is always a safe default value. Then all callers can just check exactly what they need, e.g. DivBits <= 24 if they want to optimize to a 24-bit divide.

The `AtLeast` argument should be a hint for the max _return_ value that is interesting, e.g. for a caller trying to optimize 32 bit divide to 24 bit divide they should pass in 24 (not 8). That will remove confusion about whether the value passed in should depend on IsSigned or not -- currently different callers do different things.

This can be done as an NFC cleanup before resubmitting your patch.

Reply:
Took this advice and reworked the getDivNumBits as advised. Will submit shortly after verifying tests.

Reply:
Looks like sdiv/srem failures are from catching at least one group of cases of valid shrinking where it wasn't before because of early exit from testing for -1. Thanks!